### ISSUE TITLE
sidekiq-ent limiter timeouts from redis-client revisited

### ISSUE THREAD

**Issue by 97jaz** [sidekiq/sidekiq#6159]
Ruby version: 3.3.0
Rails version: 7.0.8
Sidekiq / Pro / Enterprise version(s):  7.2.0 / 7.2.0 / 7.2.0

I believe I've figured out the actual cause of the timeout errors that we (and others) get from redis-client when using concurrent limiters.

Every time we see one of these errors, the message says something like:
```
RedisClient::ReadTimeoutError: Waited 1.0001943889074028 seconds
```
where the wait time is always 1 second + less than one millisecond.

The error stems from `sidekiq-ent` at `lib/sidekiq-ent/limiter/concurrent.rb` line 109:
```ruby
candidate = conn.blocking_call(diff + 1, "BLMOVE", @free, @pend, "RIGHT", "LEFT", diff.to_f)
```

The client is given a full second more than the server -- so you'd expect never to get a client timeout unless either the client or the server were severly over-burdened.

_But_... as it turns out, redis (or at least redis 7.0, which is what we're running) treats a timeout of less than (or equal to, it seems) 1ms as if it were 0. And a timeout of 0 means that it never times out. I've verified this:
```
irb(main):123* Sidekiq::Limiter.redis do |con|
irb(main):124*   con.blocking_call 10, "BLMOVE", "JHGJHGskhd", "nmbmbsug", "RIGHT", "LEFT", 0.00021
irb(main):125> end
/opt/rails/vendor/bundle/ruby/3.3.0/gems/redis-client-0.17.0/lib/redis_client/ruby_connection/buffered_io.rb:140:in `block in fill_buffer': Waited 10 seconds (RedisClient::ReadTimeoutError)
```
versus
```
irb(main):126* Sidekiq::Limiter.redis do |con|
irb(main):127*   con.blocking_call 10, "BLMOVE", "JHGJHGskhd", "nmbmbsug", "RIGHT", "LEFT", 0.00101
irb(main):128> end
=> nil
```

From my brief experimentation, it looks like a timeout of 0.001 is also treated as 0, but 0.00101 is not. 

Anyhow, this is why we get errors from the client. If the server timed out, then the code would raise `OverLimit`, and the job would be re-scheduled, as normal.

So I think that sidekiq-ent ought to treat a redis timeout of <= 0.001 as if it were 0.

**Comment by 97jaz** [comment]
Oh, by the way, in case you were wondering: we're _not_ using a `wait_timeout` of less than a millisecond. It's that on a subsequent iteration of the `while` loop, `diff` is sometimes that small.

**Comment by mperham** [comment]
Ok, I see:

1. A limiter call loops until the `diff` value gets down to <0.001.
2. Sidekiq issues a BLMOVE call with timeout <0.001.
3. Because Redis treats that as a zero timeout, BLMOVE times out on the client-side after 1 second.

Thank you for figuring that out!

**Comment by mperham** [comment]
7.2.3 does not display this behavior. It pauses:

```
$ redis-cli
127.0.0.1:6379> BLMOVE foo bar LEFT RIGHT 0.1
(nil)
127.0.0.1:6379> BLMOVE foo bar LEFT RIGHT 0.01
(nil)
127.0.0.1:6379> BLMOVE foo bar LEFT RIGHT 0.001
(nil)
127.0.0.1:6379> BLMOVE foo bar LEFT RIGHT 0.0001
(nil)
127.0.0.1:6379> BLMOVE foo bar LEFT RIGHT 0.000001
(nil)
127.0.0.1:6379> BLMOVE foo bar LEFT RIGHT 0
```

Whereas 0 actually blocks.

**Comment by 97jaz** [comment]
Interesting. I just verified the behavior I described above (on Redis 7.0) using `redis-cli` directly (just in case there was an issue in `redis-client`). It's blocking with a timeout of 0.001.

So this is either Redis version-dependent or possibly platform-dependent (this redis server is running on linux).

**Comment by 97jaz** [comment]
Oh, also: the exact version I'm looking at is 7.0.7.

**Comment by 97jaz** [comment]
Okay, found it: https://github.com/redis/redis/commit/574a49b96c02f6833bef3b15501aafda1e0fb031

In fact, this seems to have been fixed in the very next version (7.0.8). So that's some bad luck for us.

**Comment by mperham** [comment]
Thanks for all the research. It appears to be a Redis bug which has been fixed so I'll close with no action.
